### ISSUE TITLE
dpo_loss的疑问

### ISSUE THREAD

**Issue by zxywww** [jingyaogong/minimind#298]
```python
def dpo_loss(ref_probs, probs, beta):
    # ref_probs 和 probs 都是 shape: (batch_size, seq_len)
    # 计算每个样本的平均概率
    ref_probs = ref_probs.mean(dim=1)
    probs = probs.mean(dim=1)#log概率相加是计算assistant整个句子输出的概率, 但是为什么统一又除以seq_len?

    # 将 chosen 和 rejected 数据分开
    batch_size = ref_probs.shape[0]
    chosen_ref_probs = ref_probs[:batch_size // 2]
    reject_ref_probs = ref_probs[batch_size // 2:]
    chosen_probs = probs[:batch_size // 2]
    reject_probs = probs[batch_size // 2:]

    # 计算对数比率，比较偏好差异
    pi_logratios = chosen_probs - reject_probs
    ref_logratios = chosen_ref_probs - reject_ref_probs
    logits = pi_logratios - ref_logratios
    loss = -F.logsigmoid(beta * logits)
    return loss.mean()

```
请问 这里
'''python
 ref_probs = ref_probs.mean(dim=1)
    probs = probs.mean(dim=1)
```
每个next_token的概率log后的相加是计算assistant输出的整个句子的概率的log
但是为什么统一又除以seq_len?
不应该是求sum(dim=1)吗?

**Comment by jingyaogong** [comment]
不错的问题，这里主要就是mean和sum的区别了

严格来说对于整个序列的概率应该是各个token概率的乘积（对数概率相加），所以使用 .sum(dim=1) 

但是

平均对数概率：
1、避免长序列主导loss（因为长序列的logits总和会天然更大）
2、不同长度text对loss影响更均衡

这里说白了：

- 如果希望长序列和短序列的贡献不同，用sum
- 如果希望不同长度的样本贡献均衡，用mean

其他问题欢迎跟进，thanks

**Comment by zxywww** [comment]
您好, 谢谢解答. 但是还有一个很重要的问题就是, 如果要排除不同句子长度的影响, 也不应该直接.mean(dim=1), 因为求mean相当于每个句子都同时除以同一个数字seq_length, 但是我们应该除以的是assistant回答内容的长度, 也就是mask.sum(dim=1, keep_dim = True)

**Comment by zxywww** [comment]
> 不错的问题，这里主要就是mean和sum的区别了
> 
> 严格来说对于整个序列的概率应该是各个token概率的乘积（对数概率相加），所以使用 .sum(dim=1)
> 
> 但是
> 
> 平均对数概率： 1、避免长序列主导loss（因为长序列的logits总和会天然更大） 2、不同长度text对loss影响更均衡
> 
> 这里说白了：
> 
> * 如果希望长序列和短序列的贡献不同，用sum
> * 如果希望不同长度的样本贡献均衡，用mean
> 
> 其他问题欢迎跟进，thanks

您好, 谢谢解答. 但是还有一个很重要的问题就是, 如果要排除不同句子长度的影响, 也不应该直接.mean(dim=1), 因为求mean相当于每个句子都同时除以同一个数字seq_length, 但是我们应该除以的是assistant回答内容的长度, 也就是mask.sum(dim=1, keep_dim = True)

**Comment by zxywww** [comment]
"避免长序列主导loss（因为长序列的logits总和会天然更大）", 这句话严格说长序列的概率相乘天然更小, 对数概率相加应该是天然更小, 不是更大

**Comment by zxywww** [comment]
"避免长序列主导loss（因为长序列的logits总和会天然更大）", 这句话正确来说是: 长序列的概率连乘天然更小, 对数概率相加应该是天然更小, 不是更大

**Comment by jingyaogong** [comment]
> "避免长序列主导loss（因为长序列的logits总和会天然更大）", 这句话正确来说是: 长序列的概率连乘天然更小, 对数概率相加应该是天然更小, 不是更大

> 您好, 谢谢解答. 但是还有一个很重要的问题就是, 如果要排除不同句子长度的影响, 也不应该直接.mean(dim=1), 因为求mean相当于每个句子都同时除以同一个数字seq_length, 但是我们应该除以的是assistant回答内容的长度, 也就是mask.sum(dim=1, keep_dim = True)

是的，存在口误：`0<logits<1`连乘对数越来越小而不是越大

是的，这是严谨的做法，已更新：https://github.com/jingyaogong/minimind/commit/278ec760a191bbdb55139f67f0d3cd6138c10062

感谢指正😊

**Comment by Tallisgo** [comment]
![Image](https://github.com/user-attachments/assets/3afea77d-b0f9-49ec-a454-21bba5b8cf10)
计算loss之前已经乘过mask了，那在dpo_loss中就不需要再乘以mask了吧

```python
    seq_lengths = mask.sum(dim=1, keepdim=True)  # (batch_size, 1)
     ref_probs = (ref_probs * mask).sum(dim=1) / seq_lengths.squeeze()
     probs = (probs * mask).sum(dim=1) / seq_lengths.squeeze()
```
